skimage/filters/_sparse.py

```diff
@@ -70,28 +152,18 @@ def correlate_sparse(image, kernel, mode='reflect'):
         _validate_window_size(kernel.shape)
         padded_image = np.pad(
             image,
-            [(w//2, w//2) for w in kernel.shape],
+            [(w // 2, w // 2) for w in kernel.shape],
             mode=np_mode,
         )
-    indices = np.nonzero(kernel)
-    offsets = np.ravel_multi_index(indices, padded_image.shape).astype(np.intp)
-    values = kernel[indices].astype(padded_image.dtype)
 
-    result = np.zeros(
-        [a - b + 1 for a, b in zip(padded_image.shape, kernel.shape)],
-        dtype=padded_image.dtype,
+    indices = np.nonzero(kernel)
+    values = kernel[indices].astype(padded_image.dtype, copy=False)
+    indices = list(zip(*indices))
+    kernel_indices_and_values = [(idx, v) for idx, v in zip(indices, values)]
+    if (0, ) * kernel.ndim not in indices:
```

Comment:
```suggestion
    if kernel.reshape(-1)[0] == 0:
```

Reply:
I don't think this is equivalent? What I want is if the sparse kernel does not have an entry at the `(0,) * ndim` corner location, then an entry should be added so that there will not be a shift in the output of `_correlate_sparse`. (Internal to `_correlate_sparse`, the output array is initialized from this first point, which needs to be the corner location in order to avoid an unexpected spatial shift in the output).

See if e4cff50 makes things any clearer

Reply:
Isn't `(0,) * ndim` the index of the first entry of `kernel`? `kernel.reshape(-1)[0]` is the first entry of `kernel`. If it is `0` then `(0,) * ndim` is not in `indices`.

Reply:
Yes, I agree now that what you suggest is equivalent and would be fine with changing to that.

Reply:
On closer look, I would prefer to leave this check as-is so that it avoids a device->host transfer if using a CuPy backend. (For CuPy, `kernel.reshape(-1)[0]` will give a 0-dimensional `cupy.ndarray` rather than a host scalar and comparing to 0 requires overhead of transferring that value from the GPU back to the host to do the comparison)